sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java

```diff
@@ -107,10 +108,57 @@ public void processElement(
 
   /** Implementation of {@link #viaRandomKey()}. */
   public static class ViaRandomKey<T> extends PTransform<PCollection<T>, PCollection<T>> {
+    private boolean isHighFanoutAndLimitedInputParallelism;
+
     private ViaRandomKey() {}
 
+    /**
+     * Use a different strategy that materializes the input and prepares it to be consumed in a
+     * highly parallel fashion.
+     *
+     * <p>It is tailored to the case when input was produced in an extremely sequential way -
+     * typically by a ParDo that emits millions of outputs _per input element_, e.g., executing a
+     * large database query or a large simulation and emitting all of their results.
+     *
+     * <p>Internally, it materializes the input at a moderate cost before reshuffling it, making the
+     * reshuffling itself significantly cheaper in these extreme cases on some runners. Use this
+     * only if your benchmarks show an improvement.
+     */
+    public ViaRandomKey<T> withHintHighFanoutAndLimitedInputParallelism() {
+      this.isHighFanoutAndLimitedInputParallelism = true;
```

Comment:
Lets return a new PTransform like our other PTransform builder patterns instead of mutating the state of the current transform.